content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,173 @@
+
+Some time ago, we realized that the pipeline for one of our largest codebases was taking 45min on average to be finished. Of those, 22min running the backend tests only.
+It means that, for each PR, developers would wait around 45min to get feedback on their changes, delaying the time to deliver code to production.
+
+We have decided to create a Working Group to improve this. This post aims to share our journey to reduce the total execution time of our backend tests.
+
+## What is a Working Group?
+
+Working Groups are a great format when it comes to solve multi-squad tech debts, as they are short-lived, temporary, diverse, and they are formed to solve a cross-organization issue. Cool, right? Would you like to know more about Loadsmart Engineering culture? [Check it out](https://engineering.loadsmart.com/blog/our-engineering-culture).
+
+In this initiative, the team was formed by three full-stack engineers, two backend engineers, and one site reliability engineer. All of them from different squads.
+
+## Goals
+As with all Loadsmart initiatives, it should have clear goals and be based on data to be properly tracked by everyone.
+
+For this specific effort, we came across the following goals:
+
+- Reduce average execution time of backend tests by 50%
+    - How to measure: CircleCI backend tests step execution time
+
+- Reduce average time from code-ready to code-deployed by 50%
+    - How to measure: CircleCI backend tests step + QA deploy + Staging deploy + Production deploy
+
+
+## Methodology
+
+We followed a scientific method for problem-solving based on the following steps:
+
+- Identify a problem:
+We may have several potential issues in our code impacting the backend test step execution time.
+However, we should pick one to focus on. In this step, we took several problems and guessed which one of them would be better to start with.
+
+- Research:
+After having a clear definition of the problem, we started doing some research to understand how it was being handled by the software community.
+
+- Hypothesis:
+Based on the results of previous steps, we could gather some hypotheses in the format of "what if...".
+
+- Experimentation:
+For each hypothesis, we have evaluated the effort to test it. After qualifying it as a valid one, we prototyped it.
+We did it for a minimal set of changes enabling an initial evaluation. Our goal here was to change minimal and get the max result possible.
+
+- Results and conclusions:
+Once we had a prototype working as expected, we started gathering data. If it showed promising results, we could invest more effort in it.
+
+## Actions and Results
+
+### Baseline
+
+As a baseline, we considered the total execution time for the backend test step: 22 minutes.
+
+1. Map and split slowest Django apps
+
+We had a bunch of tests spread through several Django apps.
```

Comment:
How many Django apps?

Reply:
We have 52 Django apps 😨 

Based in your comment about exposing our vulnerabilities, I think it's better to keep this way more generic :)